internal/pkg/v2/README.md

```diff
@@ -0,0 +1,275 @@
+			support								Support-service controllers/tests
+			system								System Management controllers/tests
+		handle								Common request handling implementation
+		routing								Version, kind, action router
+```
+
+Note the separation of common cross-service code from service-specific code.
+
+
+
+### Use-case Endpoints
+
+A use-case endpoint is versioned; e.g. `/api/v2/metrics`.
+
+A use-case endpoint accepts either a single use-case request or an array of use-case requests.
+
+A single use-case request returns a single corresponding use-case response.  Success is indicated by an 200 HTTP status code.  The use-case response contains a transport-independent result code that indicates success or a specific failure reason.
+
+An array of use-case requests returns an array of corresponding use-case responses.  A 207 HTTP status code will be returned regardless of the success or failure of the individual requests.  Each use-case response contains a transport-independent result code that indicates success or a specific failure reason.
+
+Multiple use-case requests received in a single transport request must all be of same type (which corresponds to the request DTO expected by the underlying use-case).
+
+Multiple use-case requests received in a single transport request are processed concurrently.  Results are aggregated and returned as a single array of responses in the transport's response.  The relative position of a use-case response in the response's array has no correlation to the relative position of the corresponding use-case request in the request's array.
+
+
+
+### Batch Endpoint
+
+There is only one batch endpoint and it is unversioned: `/api/batch`.  
+
+A RequestEnvelope DTO and a ResponseEnvelope DTO wrap use-case request DTOs and use-case response DTOs respectively for requests and response sent via the batch endpoint.  These batch-specific DTOs contain version, type, and action properties.  They also include a strategy and a content property.  The content property contains a use-case-specific request or response that corresponds to the version, type, and action in the envelope.
+
+A batch endpoint accepts an array of wrapped use-case requests and returns an array of wrapped corresponding use-case responses.  A 207 HTTP status code will be returned regardless of the success or failure of the individual requests.  Each use-case response contains a transport-independent result code that indicates success or a specific failure reason.  
+
+Multiple wrapped use-case requests received in a single transport request may be of mixed types.
+
+Multiple wrapped use-case requests received in a single transport request are processed sequentially.  Results are aggregated and returned as a single array of wrapped responses in the transport's response. The relative position of a wrapped use-case response in the response's array has no correlation to the relative position of the corresponding wrapped use-case request in the request's array.
```

Comment:
> Add why

It was a somewhat arbitrary decision.  There's forward vision that would provide for the array of RequestEnvelope DTOs to be wrapped in an object that specified a processing strategy.  This would allow for much more complex saga-like interaction (i.e. validate that the array contains all of the items necessary to do some complex behavior; process all of the objects of type A concurrently; wait for type A processing to complete; process type B, type C, and type D objects sequentially and in that order; return a result).  You could also implement a strategy that provided for simple DTO-ignorant concurrent processing of all RequestEnvelope DTOs.

> Might want to add how the responses are match to the requests, i.e. requestId

IMO, how requests and responses are correlated is an issue for the OpenAPI specification.

Unresolve and comment if you want to continue this conversation here.

Reply:
You gave me the background on the Why? But didn't address it in the document.

Reply:
Since you bring up the correlation issues, i.e. response's array has no correlation to the relative position it begs the question on how correlation is resolved.

Reply:
> You gave me the background on the Why? But didn't address it in the document.

I thought I did; it's in the document's "future" section.

![image](https://user-images.githubusercontent.com/464327/74238128-c27f0400-4c9a-11ea-9ecc-d27137c1f6ed.png)

> Since you bring up the correlation issues, i.e. response's array has no correlation to the relative position it begs the question on how correlation is resolved.

"IMO, how requests and responses are correlated is an issue for the OpenAPI specification."  It's a data/protocol-level concern and is addressed in the OpenAPI spec.

https://github.com/edgexfoundry/edgex-go/pull/2285#discussion_r376386873